File: backend/app/integrations/stubs.py
"""Deterministic stub providers used until real adapters are wired.

Stubs are seeded by input so responses are stable across calls — good
enough for demos and tests, and honest about being synthetic (every
response carries ``"source": "stub"``).
"""

import hashlib
from typing import Any

from app.utils import utcnow_iso

_RISKS = ["low", "medium", "high"]
_TRENDS = ["falling", "stable", "rising"]


def _bucket(key: str, size: int) -> int:
    digest = hashlib.sha256(key.encode()).digest()
    return digest[0] % size


class StubWeatherProvider:
    def get_forecast(self, island: str) -> dict[str, Any]:
        risk = _RISKS[_bucket(f"weather:{island}", len(_RISKS))]
        return {
            "source": "stub",
            "island": island,
            "risk": risk,
            "summary": f"{risk} storm risk over the next 72h",
            "observed_at": utcnow_iso(),
        }

    def get_active_alerts(self) -> list[dict[str, Any]]:
        return []


class StubRoutingProvider:
    def estimate_transit(self, origin_island: str, destination_island: str) -> dict[str, Any]:
        if origin_island == destination_island:
            hours = 4
            mode = "land"
        else:
            hours = 12 + _bucket(f"route:{origin_island}->{destination_island}", 36)
            mode = "sea"
        return {
            "source": "stub",
            "origin": origin_island,
            "destination": destination_island,
            "mode": mode,
            "transit_hours": hours,
        }


class StubPricingProvider:
    def get_price(self, crop_name: str, island: str | None = None) -> dict[str, Any]:
        base = 1.0 + _bucket(f"price:{crop_name.lower()}", 40) / 10.0
        trend = _TRENDS[_bucket(f"trend:{crop_name.lower()}:{island}", len(_TRENDS))]
        return {
            "source": "stub",
            "crop_name": crop_name,
            "island": island,
            "unit_price_usd": round(base, 2),
            "trend": trend,
        }

File: src/lib/server/integrations/stubs.ts
/**
 * Deterministic stub providers used until real adapters are wired.
 *
 * Stubs are seeded by input so responses are stable across calls — good
 * enough for demos and tests, and honest about being synthetic (every
 * response carries `"source": "stub"`).
 */

import { createHash } from "node:crypto";

import { utcnowIso } from "../time";
import type {
  CropPrice,
  PricingProvider,
  RoutingProvider,
  TransitEstimate,
  WeatherAlert,
  WeatherForecast,
  WeatherProvider,
} from "./interfaces";

const RISKS = ["low", "medium", "high"] as const;
const TRENDS = ["falling", "stable", "rising"] as const;

function bucket(key: string, size: number): number {
  return createHash("sha256").update(key).digest()[0] % size;
}

export class StubWeatherProvider implements WeatherProvider {
  async getForecast(island: string): Promise<WeatherForecast> {
    const risk = RISKS[bucket(`weather:${island}`, RISKS.length)];
    return {
      source: "stub",
      island,
      risk,
      summary: `${risk} storm risk over the next 72h`,
      observed_at: utcnowIso(),
    };
  }

  async getActiveAlerts(): Promise<WeatherAlert[]> {
    return [];
  }
}

export class StubRoutingProvider implements RoutingProvider {
  estimateTransit(originIsland: string, destinationIsland: string): TransitEstimate {
    const sameIsland = originIsland === destinationIsland;
    return {
      source: "stub",
      origin: originIsland,
      destination: destinationIsland,
      mode: sameIsland ? "land" : "sea",
      transit_hours: sameIsland
        ? 4
        : 12 + bucket(`route:${originIsland}->${destinationIsland}`, 36),
    };
  }
}

export class StubPricingProvider implements PricingProvider {
  getPrice(cropName: string, island: string | null = null): CropPrice {
    const base = 1.0 + bucket(`price:${cropName.toLowerCase()}`, 40) / 10.0;
    const trend = TRENDS[bucket(`trend:${cropName.toLowerCase()}:${island}`, TRENDS.length)];
    return {
      source: "stub",
      crop_name: cropName,
      island,
      unit_price_usd: Math.round(base * 100) / 100,
      trend,
    };
  }
}
